manager/allocator/network/ipam/ipam.go

```diff
@@ -0,0 +1,857 @@
+// NetworkAttachment (and an error). This means it does not mutate its
+// argument.
+func (a *allocator) AllocateAttachment(config *api.NetworkAttachmentConfig) (attach *api.NetworkAttachment, rerr error) {
+	// first, get the network
+	// this ONLY WORKS because a network cannot be updated. if a network can be
+	// updated, we'd have a lot more headaches.
+	local, ok := a.networks[config.Target]
+	if !ok {
+		return nil, errors.ErrDependencyNotAllocated("network", config.Target)
+	}
+
+	attachment := &api.NetworkAttachment{
+		Network:              local.nw,
+		Aliases:              config.Aliases,
+		DriverAttachmentOpts: config.DriverAttachmentOpts,
+		Addresses:            make([]string, 0, len(config.Addresses)),
+	}
+
+	// now get the IPAM driver and options
+	ipam, _ := a.drvRegistry.IPAM(local.nw.IPAM.Driver.Name)
+	ipamOpts := local.nw.IPAM.Driver.Options
+
+	// there are two cases. the most common case is that no addresses are
+	// specified. This is the case for most tasks. The second case is when some
+	// addresses are specified in the config. This is only the case for
+	// attachments belonging to a NetworkAttachment tasks, which are used to
+	// connect a regular docker container to an attachable networks.
+	if len(config.Addresses) == 0 {
+		// we need one address for the attachment on the network. the network
+		// may have many disjoint IP address pools, and some of them may be
+		// exhausted of IP addresses. So, we have to try allocating on each
+		// pool belonging to the network until we get one that succeeds
+		for _, poolID := range local.pools {
+			// passing nil as the 2nd argument tells the IPAM driver that we're
+			// requesting no particular address, and to give us whatever is
+			// available.
+			ip, _, err := ipam.RequestAddress(poolID, nil, ipamOpts)
+			if err == nil {
+				// keep track of which pool this address belongs to, for
+				// deallocation
+				local.endpoints[ip.String()] = poolID
+				attachment.Addresses = append(attachment.Addresses, ip.String())
+				return attachment, nil
+			}
+			// if we get ErrNoAvailableIPs, that means we should try the next
+			// pool. any other error is fatal.
+			if err != ipamapi.ErrNoAvailableIPs {
+				return nil, errors.ErrInternal("error requesting address from ipam: %v", err)
+			}
+		}
+		// if we get all the way through the loop of pools without returning
+		// either an attachment or an error, that means the address space for
+		// the network is exhausted
+		return nil, errors.ErrResourceExhausted("ip addresses", "address space for network %v is exhausted", local.nw.ID)
+	}
+	// otherwise, if there are addresses specified, this is a NetworkAttachment
+	// task, and we need to allocate those specific addresses
+
+	// if the allocation of any of these addresses fails, they all need to be
+	// released
+	defer func() {
+		if rerr != nil {
+			for _, addr := range attach.Addresses {
+				poolID := local.endpoints[addr]
```

Comment:
Do we need to delete it from `local.endpoints[addr]` as well?

Reply:
good catch.

Reply:
This comment has been addressed.